### PR TITLE
Added hash functions from libpromises and added StringCopyTruncateAndHashIfNecessary()

### PR DESCRIPTION
**Pull request by olehermanse**

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a PR! Maybe @vpodzime can review this?

File: libutils/hash.c [hunk 1]
```diff
@@ -674,3 +674,56 @@ char *SkipHashType(char *hash)
...
+    char *const hash_prefix_start = terminator - md5_string_length;
+    assert(hash_prefix_start >= dst);
+    char *const hash_start = hash_prefix_start + sizeof(md5_prefix) - 1;
+    assert(hash_start > hash_prefix_start);
+
+    // Insert the "#MD5=" part into dst;
+    memcpy(hash_prefix_start, md5_prefix, md5_prefix_length);
+
+    // Produce the hex string representation of MD5 hash:
+    // (Overwrite the last part of dst)
```
Comment (vpodzime) [inline libutils/hash.c]: Maybe we should have a function for this? Or, better, can't it use the `HashPrintSafe` function from the previous commit?

File: libutils/hash.c [hunk 2]
```diff
@@ -394,3 +396,281 @@ HashSize HashSizeFromId(HashMethod hash_id)
...
+    const int len,
+    unsigned char digest[EVP_MAX_MD_SIZE + 1],
+    HashMethod type)
+{
+    assert(buffer != NULL);
+    assert(digest != NULL);
+
+    memset(digest, 0, EVP_MAX_MD_SIZE + 1);
+
+    if (type == HASH_METHOD_CRYPT)
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/hash.c]: I guess, but I didn't change any code in this commit. Just moving it from core.

File: libutils/hash.c [hunk 3]
```diff
@@ -394,3 +396,281 @@ HashSize HashSizeFromId(HashMethod hash_id)
...
+        StringBytesToHex(&dst[dst_len], dst_size - dst_len,
+                         digest, digest_len);
+    assert(ret == 2 * digest_len);
+
+#if 0 /* TODO return proper exit status and check it in the callers */
+    if (ret < 2 * digest_len)
+    {
+        return NULL;
+    }
+#endif
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/hash.c]: I think we should clean up this whole function in a separate PR.

File: libutils/hash.c [hunk 4]
```diff
@@ -394,3 +396,281 @@ HashSize HashSizeFromId(HashMethod hash_id)
...
+        Log(LOG_LEVEL_ERR, "Failed to allocate openssl hashing context");
+        return;
+    }
+
+    if (EVP_DigestInit(context, md) == 1)
+    {
+        EVP_DigestUpdate(context, buffer, len);
+
+        unsigned int digest_length;
+        EVP_DigestFinal(context, digest, &digest_length);
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/hash.c]: Done.

File: libutils/hash.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/hash.c]: Added assert above.

File: libutils/hash.c [hunk 5]
```diff
@@ -394,3 +396,281 @@ HashSize HashSizeFromId(HashMethod hash_id)
...
+    assert(key != NULL);
+
+    memset(digest, 0, EVP_MAX_MD_SIZE + 1);
+
+    if (type == HASH_METHOD_CRYPT)
+    {
+        Log(LOG_LEVEL_ERR,
+            "The crypt support is not presently implemented, please use sha256 instead");
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/hash.c]: Added assert above.

File: libutils/hash.c [hunk 6]
```diff
@@ -394,3 +396,281 @@ HashSize HashSizeFromId(HashMethod hash_id)
...
+        return false;
+    }
+
+    for (int i = 0; i < size; i++)
+    {
+        if (digest1[i] != digest2[i])
+        {
+            return false;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/hash.c]: Added.

File: libutils/hash.c (hunk 1 above)
Comment (olehermanse) [reply libutils/hash.c]: Let's get this merged and working first, then I can do refactoring / smaller changes like this.

Comment (olehermanse) [reply libutils/hash.c]: For context, I didn't write this part. This function replaces `GenerateMd5Hash()` in `locks.c`.

**Pull request merged**